src/openmmio.py

```diff
@@ -1117,34 +1168,68 @@ def normal_modes(self, shot=0, optimize=True):
         normal_modes = normal_modes[larger_freq_idxs]
         return freqs, normal_modes
 
-    def optimize(self, shot=0, crit=1e-4):
+    def optimize(self, shot, crit=1e-4, disable_vsite=False, align=True, include_restraint_energy=False):
+
+        """ 
+        Optimize the geometry and align the optimized 
+        geometry to the starting geometry.
+        
+        Parameters
+        ----------
+        shot : int
+            The snapshot number to be minimized
+        crit : float
+            Convergence criterion in kJ/mol
+        disable_vsite : bool
+            Disable virtual sites (needed for SMIRNOFF)
+        include_restraint_energy : bool
+            Include energy component from CustomExternalForce
 
-        """ Optimize the geometry and align the optimized geometry to the starting geometry, and return the RMSD. """
+        Returns
+        -------
+        E : float
+            Potential energy of the system
+        rmsd : float
+            RMSD of the system (w/r.t. starting geometry) in Angstrom
+        """
 
         steps = int(max(1, -1*np.log10(crit)))
         self.update_simulation()
         self.set_positions(shot)
+        if hasattr(self, 'restrain_k'):
+            self.set_restraint_positions(shot)
         # Get the previous geometry.
         X0 = np.array([j for i, j in enumerate(self.simulation.context.getState(getPositions=True).getPositions().value_in_unit(angstrom)) if self.AtomMask[i]])
+        # printcool_dictionary(energy_components(self.simulation), title='Energy component analysis before minimization, shot %i' % shot)
         # Minimize the energy.  Optimizer works best in "steps".
         for logc in np.linspace(0, np.log10(crit), steps):
-            self.simulation.minimizeEnergy(tolerance=10**logc*kilojoule/mole)
+            self.simulation.minimizeEnergy(tolerance=10**logc*kilojoule/mole, maxIterations=10000)
+        # Remove the restraint energy from the total energy if desired.
+        groups = set(range(32))
+        if hasattr(self, 'restrain_k') and not include_restraint_energy:
```

Comment:
Same as above, we may want to check `if hasattr(self, 'restrain_k') and self.restrain_k != 0`

Reply:
I thought having this line everywhere was cumbersome so instead I check if self.restraint_frc_index is not None.